client/src/main/java/io/pravega/client/stream/impl/SegmentTransactionImpl.java

```diff
@@ -42,15 +83,22 @@ public UUID getId() {
 
     @Override
     public void flush() throws TxnFailedException {
+        checkFailed();
         try {
             out.flush();
+            synchronized (lock) {
+                removeCompleted();
+                checkFailed();
+                Preconditions.checkState(outstanding.isEmpty());
+            }
         } catch (SegmentSealedException e) {
             throw new TxnFailedException(e);
         }
     }
 
     @Override
     public void close() throws TxnFailedException {
+        flush();
         try {
             out.close();
```

Comment:
Based on the below two points I suspect that invoking `flush()` inside `close()` would cause an issue
- out.close() also invokes a flush internally. 
- also NoSuchSegment response would have closed the state of the SegmentOutputStreamImpl  --> hence out.close() will not invoke out.flush() if the state is closed.

Reply:
It's not an issue. I'm invoking the flush() here to get the logic above to execute in addition to the normal flushing. The idea being I don't want close() to complete successfully if one of the inflight events fails. Calling out.flush() twice which will occur in this case is a no-op as when there are no inflight events it just returns. Similarly if close() has already been invoked on the output stream, either from segment does not exist or something else it is a no-op and returns. So while flush() and close() may be called more than once, these operations are idempotent.